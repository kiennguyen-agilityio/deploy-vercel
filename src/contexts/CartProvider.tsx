import { createContext, useMemo, useCallback, ReactNode, useReducer, useContext } from 'react';
import { mutate } from 'swr';

// constants
import {
  ADD_PRODUCT_FAILURE,
  DELETE_PRODUCT_FAILURE,
  UPDATE_PRODUCT_FAILURE,
  ADD_PRODUCT_REQUEST,
  ADD_PRODUCT_SUCCESS,
  DELETE_PRODUCT_REQUEST,
  DELETE_PRODUCT_SUCCESS,
  UPDATE_PRODUCT_REQUEST,
  UPDATE_PRODUCT_SUCCESS,
} from '@/constants/actionType';

// services
import { addToCart, deleteCartItem, updateItem } from '@/services/api-action';

// hooks
import { useCartProducts } from '@/components/hooks/fetch';

// constants
import { CART_API } from '@/constants/api';

// reducers
import cartReducer, { ProductCartState } from '@/reducers/cartReducer';

// interfaces
import Product from '@/interface/Product';

export const initialState: ProductCartState = {
  error: '',
  isLoading: false,
  isAdding: false,
  products: [],
};

export interface CartContextProps {
  error: string;
  isAdding: boolean;
  isLoading: boolean;
  products?: Product[];
  onAddToCart: (product: Product) => void;
  onUpdateCart: (productId: string, updatedData: Partial<Product>, url: string) => Promise<void>;
  onDeleteProductFromCart: (productId: string) => void;
}

export const CartContext = createContext<CartContextProps>({
  error: '',
  isAdding: false,
  isLoading: false,
  products: [],
  onAddToCart: () => undefined,
  onUpdateCart: async () => {},
  onDeleteProductFromCart: async () => {},
});

interface CartProviderProps {
  children: ReactNode;
}

export const useProductCartContext = () => useContext(CartContext);

const CartProvider: React.FC<CartProviderProps> = ({ children }) => {
  const [state, dispatch] = useReducer(cartReducer, initialState);
  const { isLoading, error, isAdding } = state;
  const { data: productsFromSwr } = useCartProducts();

  const handleAddToCart = useCallback(async (product: Product): Promise<void> => {
    try {
      dispatch({ type: ADD_PRODUCT_REQUEST });

      const data = await addToCart(product);

      mutate(CART_API);

      dispatch({ type: ADD_PRODUCT_SUCCESS, payload: data });
    } catch {
      dispatch({ type: ADD_PRODUCT_FAILURE });
    }
  }, []);

  const handleDeleteProductFromCart = useCallback(async (productId: string): Promise<void> => {
    try {
      dispatch({ type: DELETE_PRODUCT_REQUEST });

      await deleteCartItem(productId);

      mutate(CART_API);

      dispatch({ type: DELETE_PRODUCT_SUCCESS, payload: productId });
    } catch (error) {
      dispatch({ type: DELETE_PRODUCT_FAILURE });
    }
  }, []);

  const handleUpdateItem = useCallback(
    async (productId: string, updatedData: Partial<Product>, url: string) => {
      try {
        dispatch({ type: UPDATE_PRODUCT_REQUEST });
        await updateItem(productId, updatedData, url);

        mutate(CART_API);

        const updatedProductData: Partial<Product> = {
          ...updatedData,
          id: productId,
        };

        dispatch({ type: UPDATE_PRODUCT_SUCCESS, payload: updatedProductData });
      } catch {
        dispatch({ type: UPDATE_PRODUCT_FAILURE });
      }
    },
    []
  );

  const cartContextValue = useMemo(
    () => ({
      isLoading,
      error,
      products: productsFromSwr || [],
      isAdding,
      onAddToCart: handleAddToCart,
      onUpdateCart: handleUpdateItem,
      onDeleteProductFromCart: handleDeleteProductFromCart,
    }),
    [state, productsFromSwr]
  );

  return <CartContext.Provider value={cartContextValue}>{children}</CartContext.Provider>;
};

export default CartProvider;
