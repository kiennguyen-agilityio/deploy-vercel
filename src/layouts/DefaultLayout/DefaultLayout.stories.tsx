import { StoryObj, Meta } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import DefaultLayout from './DefaultLayout';

export default {
  title: 'Layouts/DefaultLayout',
  component: DefaultLayout,
  decorators: [
    (Story) => (
      <MemoryRouter>
        <Story />
      </MemoryRouter>
    ),
  ],
} as Meta;

type Story = StoryObj<typeof DefaultLayout>;

export const Default: Story = {
  args: {
    children: <div>This is the main content</div>,
  },
};
