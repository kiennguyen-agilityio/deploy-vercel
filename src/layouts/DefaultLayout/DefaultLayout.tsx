import { ReactNode, memo } from 'react';

// layouts
import Footer from '@/layouts/Footer/Footer';
import Header from '@/layouts/Header/Header';

interface Props {
  children: ReactNode;
}

const DefaultLayout: React.FC<Props> = ({ children }) => {
  return (
    <div className="container">
      <Header />
      <main className="main-content">{children}</main>
      <Footer />
    </div>
  );
};

export default memo(DefaultLayout);
