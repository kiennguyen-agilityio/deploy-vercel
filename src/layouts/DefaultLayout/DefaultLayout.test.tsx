import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import CartProvider from '@/contexts/CartProvider';
import DefaultLayout from './DefaultLayout';

test('renders correctly with children', () => {
  const { container } = render(
    <MemoryRouter>
      <CartProvider>
        <DefaultLayout>
          <div>Child component</div>
        </DefaultLayout>
      </CartProvider>
    </MemoryRouter>
  );
  expect(container.firstChild).toMatchSnapshot();
});
