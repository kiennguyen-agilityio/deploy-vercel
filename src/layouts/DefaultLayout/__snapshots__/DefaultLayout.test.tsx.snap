// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with children 1`] = `
<div
  class="container"
>
  <div
    class="navbar"
  >
    <a
      href="/"
    >
      <div
        class="logo"
      >
        <i
          class="fa-sharp fa-solid fa-bag-shopping"
        />
        <span
          class="shop-name"
        >
          Shop Bags
        </span>
      </div>
    </a>
    <div
      class="search-bar-container"
    >
      <input
        class="search-bar"
        value=""
      />
      <i
        class="search-icon fa-sharp fa-solid fa-search"
      />
    </div>
    <div
      class="icons"
    >
      <a
        class="cart-icon-container"
        href="/cart"
      >
        <i
          class="material-symbols-outlined icon"
        >
          shopping_cart
        </i>
      </a>
      <i
        class="material-symbols-outlined icon"
      >
        favorite
      </i>
    </div>
    <div
      class="button-container"
    >
      <button
        class="btn-sign-up"
        type="button"
      >
        Sign Up
      </button>
      <button
        class="btn-login"
        type="button"
      >
        Login
      </button>
    </div>
  </div>
  <main
    class="main-content"
  >
    <div>
      Child component
    </div>
  </main>
  <footer
    class="footer-distributed"
  >
    <div
      class="logo-footer"
    >
      <i
        class="fa-sharp fa-solid fa-bag-shopping"
      />
      <span
        class="shop-name"
      >
        Shop Bags
      </span>
    </div>
    <div
      class="social-icons"
    >
      <a
        href="https://www.facebook.com"
      >
        <i
          class="fa-brands fa-square-facebook"
        />
      </a>
      <a
        href="https://www.twitter.com"
      >
        <i
          class="fa-brands fa-square-twitter"
        />
      </a>
      <a
        href="https://www.linkedin.com"
      >
        <i
          class="fa-brands fa-linkedin"
        />
      </a>
      <a
        href="https://www.telegram.org"
      >
        <i
          class="fa-brands fa-square-instagram"
        />
      </a>
    </div>
    <p
      class="copyright"
    >
      2019 Â© Shopbag - All right reserved
    </p>
  </footer>
</div>
`;
